Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/recovery/core/RecoveryCertificateRepository.kt

```diff
@@ -1,42 +1,108 @@
 import de.rki.coronawarnapp.util.coroutine.AppScope
 import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.flow.HotDataFlow
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.flow.Flow
-import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.last
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.onStart
+import kotlinx.coroutines.flow.transform
+import kotlinx.coroutines.plus
 import timber.log.Timber
 import javax.inject.Inject
 import javax.inject.Singleton
 
 @Singleton
 class RecoveryCertificateRepository @Inject constructor(
     @AppScope private val appScope: CoroutineScope,
-    private val dispatcherProvider: DispatcherProvider,
+    dispatcherProvider: DispatcherProvider,
     private val qrCodeExtractor: DccQrCodeExtractor,
-    valueSetsRepository: ValueSetsRepository,
+    private val storage: RecoveryCertificateStorage,
 ) {
 
-    val certificates: Flow<Set<RecoveryCertificateWrapper>> = flowOf(emptySet())
+    private val internalData: HotDataFlow<Set<RecoveryCertificateContainer>> = HotDataFlow(
+        loggingTag = TAG,
+        scope = appScope + dispatcherProvider.IO,
+        sharingBehavior = SharingStarted.Lazily,
+    ) {
+        storage.recoveryCertificates
+            .map { recoveryCertificate ->
+                RecoveryCertificateContainer(
+                    data = recoveryCertificate,
+                    qrCodeExtractor = qrCodeExtractor
+                )
+            }
+            .toSet()
+            .also { Timber.tag(TAG).v("Restored recovery certificate data: %s", it) }
+    }
+
+    init {
+        internalData.data
+            .onStart { Timber.tag(TAG).d("Observing data.") }
+            .onEach { recoveryCertificates ->
+                Timber.tag(TAG).v("Recovery Certificate data changed: %s", recoveryCertificates)
+                storage.recoveryCertificates = recoveryCertificates.map { it.data }.toSet()
+            }
+            .catch {
+                it.reportProblem(TAG, "Failed to snapshot recovery certificate data to storage.")
+                throw it
+            }
+            .launchIn(appScope + dispatcherProvider.IO)
+    }
 
+    val certificates: Flow<Set<RecoveryCertificateWrapper>> =
+        internalData.data.transform { set ->
+            set.map { RecoveryCertificateWrapper(null, it) }.toSet()
+        }
+
+    @Throws(DuplicateRecoveryCertificateException::class)
     suspend fun requestCertificate(qrCode: RecoveryCertificateQRCode): RecoveryCertificateContainer {
         Timber.tag(TAG).d("requestCertificate(qrCode=%s)", qrCode)
-        throw NotImplementedError()
+        qrCodeExtractor.extract(qrCode.qrCode).toContainer().apply {
+            if (internalData.data.last().contains(this)) {
```

Comment:
This hangs forever. `last()` waits for the terminal operator of a flow, but `internalData.data` never completes in normal operation. You likely want `first()` here to get the newest value.

Reply:
But then we are left with a racecondition. Between performing the check here, and updating the internal data with `updateBlocking`, the data may have changed and we could be adding a duplicate certificate. Unlikely, but still...

Place the check inside `updateBlocking { currentData -> .... }` and perform it against `currentData` to prevent the racecondition.